Sources/Workspace/Workspace.swift

```diff
@@ -2000,6 +2021,79 @@ extension Workspace {
+
+            do {
+                try fileSystem.createDirectory(destinationDirectory, recursive: true)
+                if fileSystem.exists(tempExtractionDirectory) {
+                    try fileSystem.removeFileTree(tempExtractionDirectory)
+                }
+                try fileSystem.createDirectory(tempExtractionDirectory, recursive: true)
+            } catch {
+                diagnostics.emit(error)
+            }
+
+            group.enter()
+            self.archiver.extract(from: artifact.path, to: tempExtractionDirectory, completion: { extractResult in
+                defer { group.leave() }
+
+                switch extractResult {
+                case .success:
+                    var artifactPath: AbsolutePath? = nil
+                    diagnostics.wrap {
+                        // copy from temp location to actual location
+                        let content = try self.fileSystem.getDirectoryContents(tempExtractionDirectory)
+                        for file in content {
+                            let source = tempExtractionDirectory.appending(component: file)
+                            let destination = destinationDirectory.appending(component: file)
+                            if self.fileSystem.exists(destination) {
+                                try self.fileSystem.removeFileTree(destination)
+                            }
+                            try self.fileSystem.copy(from: source, to: destination)
+                            if destination.basenameWithoutExt == artifact.targetName {
+                                artifactPath = destination
+                            }
+                        }
+                        // remove temp location
+                        try self.fileSystem.removeFileTree(tempExtractionDirectory)
+                    }
+
+                    guard let mainArtifactPath = artifactPath else {
+                        return diagnostics.emit(.localArtifactNotFound(targetName: artifact.targetName, artifactName: artifact.targetName))
+                    }
+
+                    result.append(
+                        .local(
+                            packageRef: artifact.packageRef,
+                            targetName: artifact.targetName,
+                            path: mainArtifactPath,
+                            checksum: self.checksum(forBinaryArtifactAt: artifact.path, diagnostics: diagnostics)
+                        )
+                    )
+
+                case .failure(let error):
+                    let reason = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
+
+                    diagnostics.emit(.localArtifactFailedExtraction(artifactPath: artifact.path, targetName: artifact.targetName, reason: reason))
+                }
+            })
+        }
+
+        group.wait()
+
+        return result.map{ $0 }
+    }
+
+    private func isAtArtifactsDirectory(_ artifact: ManagedArtifact) -> Bool {
```

Comment:
nit: this can be a private extension on ManagedArtifact

Reply:
Right, on the one hand, it belongs there, but on the other, we would have to provide the artifacts directory each time making the call site a bit more verbose, i.e: `existingArtifact.is(at: self.location.artifactsDirectory)`, would that be fine?

Reply:
your call!

Reply:
Then I would leave it as is :)